python/ray/data/datasource/databricks_uc_datasource.py

```diff
@@ -0,0 +1,188 @@
+import json
+import logging
+import time
+from typing import List, Optional
+from urllib.parse import urlencode, urljoin
+
+import pyarrow
+import requests
+
+from ray.data.block import BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask
+from ray.util.annotations import PublicAPI
+
+logger = logging.getLogger(__name__)
+
+
+class _DatabricksUCReader(Reader):
+    def __init__(
+        self,
+        host: str,
+        token: str,
+        warehouse_id: str,
+        catalog: str,
+        schema: str,
+        query: str,
+    ):
+        self.host = host
+        self.token = token
+        self.warehouse_id = warehouse_id
+        self.catalog = catalog
+        self.schema = schema
+        self.query = query
+
+        url_base = f"https://{self.host}/api/2.0/sql/statements/"
+
+        payload = json.dumps(
+            {
+                "statement": self.query,
+                "warehouse_id": self.warehouse_id,
+                "wait_timeout": "0s",
+                "disposition": "EXTERNAL_LINKS",
+                "format": "ARROW_STREAM",
+                "catalog": self.catalog,
+                "schema": self.schema,
+            }
+        )
+
+        req_auth = ("token", self.token)
+        req_headers = {"Content-Type": "application/json"}
+
+        response = requests.post(
+            url_base,
+            auth=("token", self.token),
+            headers={"Content-Type": "application/json"},
+            data=payload,
+        )
+        response.raise_for_status()
+        statement_id = response.json()["statement_id"]
+
+        state = response.json()["status"]["state"]
+
+        logger.info(f"Waiting for query {query} exeuction result.")
+        try:
+            while state in ["PENDING", "RUNNING"]:
+                time.sleep(1)
+                response = requests.get(
+                    urljoin(url_base, statement_id) + "/",
+                    auth=req_auth,
+                    headers=req_headers,
+                )
+                response.raise_for_status()
+                state = response.json()["status"]["state"]
+        except KeyboardInterrupt:
+            # User cancel the command, so we cancel query execution.
+            requests.post(
+                urljoin(url_base, f"{statement_id}/cancel"),
+                auth=req_auth,
+                headers=req_headers,
+            )
+            try:
+                response.raise_for_status()
+            except Exception as e:
+                logger.warning(
+                    f"Canceling query {query} execution failed, reason: {repr(e)}."
+                )
+            raise
+
+        if state != "SUCCEEDED":
+            raise RuntimeError(f"Query {self.query} execution failed.")
```

Comment:
Here and elsewhere so that the query is distinct from the error message
```suggestion
            raise RuntimeError(f"Query {self.query!r} execution failed.")
```